sstables/mx/reader.cc

```diff
@@ -1612,6 +1623,15 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
                     }
                     _index_in_current_partition = false;
                     _read_enabled = false;
+                    if (_index_reader->eof()) {
+                        // Close the SSTable reader proactively, if the index is completely exhausted
+                        // and no partition was found in the current fast-forward call. This allows
+                        // disk space of SSTables to be reclaimed earlier if they take part in a
+                        // long-living read and they're deleted midway.
+                        sstlog.trace("Closing reader {} for {} after fast-forward call found that index reached EOF and there's nothing left to read",
+                                     fmt::ptr(this), _sst->get_filename());
+                        return close();
+                    }
```

Comment:
This really frightens me. Let's see what tests show.

Reply:
I am convinced it works, but I'd also like to have something that will fit more naturally with the flow. A reader is only closed once destroyed by whoever keeps it. I share your feeling and have been giving this some thought.

I wouldn't like to penalize others, with always having an additional layer of indirection.

So what about wrapping it only for cleanup with a 

auto_closed_sstable_reader, which stores the sstable partition range and the sstable reader itself. any request is forwarded to the underlying reader

if fast forward is called beyond the range spanned by the sstable, we can safely close the underlying sstable reader and replace it with an empty reader.

Much better, right?

Reply:
I think it looks good: https://github.com/raphaelsc/scylla/commit/a94ecd422f2b470c977b1efba7320932d468cbf3#diff-79da4c6195a1867e40a679218fb17f5e308e433983d351da9adbdcf32271e341R1227